File: epg-generator.php
<?php

// 待處理的EPG來源
$sources = [
    "epgshanghai.xml" => [
        "url" => "https://epg.deny.vip/sh/tel-epg.xml",
        "id_suffix" => "shanghai"  // 用於拼接頻道ID的後綴
    ],

"heiptv.xml" => [
        "url" => "https://epg.mb6.top/heiptv.xml",
 "id_suffix" => "hebeiiptv"  // 用於拼接頻道ID的後綴
    ],
    
    
    "epg51zmt.xml" => [
        "url" => "https://epg.51zmt.top:8001/difang.xml",
        "id_suffix" => "51zmt"     // 用於拼接頻道ID的後綴
    ],
    "epgunifi.xml" => [         // 沒有後綴的情況
        "url" => "https://raw.githubusercontent.com/AqFad2811/epg/refs/heads/main/unifitv.xml",
        // 沒有 id_suffix 欄位
    ]
];

$maxWorkers = 4; // 最大並行處理進程數

/**
 * 取得頻道顯示名稱（優先中文）
 */
function get_display_name($channel) {
    // 尋找中文名稱
    $xpath = new DOMXPath($channel->ownerDocument);
    $zhName = $xpath->query(".//display-name[@lang='zh']", $channel);
    if ($zhName->length > 0 && trim($zhName->item(0)->nodeValue) !== '') {
        return trim($zhName->item(0)->nodeValue);
    }
    
    // 尋找任意語言名稱
    $allNames = $xpath->query(".//display-name", $channel);
    foreach ($allNames as $node) {
        if (trim($node->nodeValue) !== '') {
            return trim($node->nodeValue);
        }
    }
    
    // 嘗試從頻道ID取得
    if ($channel->hasAttribute('id') && $channel->getAttribute('id') !== '') {
        return $channel->getAttribute('id');
    }
    
    return "未知頻道";
}

/**
 * 處理XML內容
 */
function process_xml($xmlContent, $idSuffix) {
    if (empty($xmlContent)) return null;
    
    try {
        $dom = new DOMDocument();
        $dom->preserveWhiteSpace = false;
        $dom->formatOutput = true;
        $dom->loadXML($xmlContent, LIBXML_NOERROR | LIBXML_NOWARNING);
        
        $channelMap = [];
        $xpath = new DOMXPath($dom);
        
        // 處理所有頻道節點
        $channels = $xpath->query('//channel');
        foreach ($channels as $channel) {
            $oldId = $channel->getAttribute('id');
            $displayName = get_display_name($channel);
            
            // 檢查是否需要新增後綴
            if (!empty($idSuffix)) {
                // 使用id_suffix拼接新的頻道ID
                $newId = ($oldId !== $displayName) 
                    ? "{$displayName}_{$idSuffix}"
                    : "{$oldId}_{$idSuffix}";
            } else {
                // 不新增後綴
                $newId = ($oldId !== $displayName) 
                    ? $displayName
                    : $oldId;
            }
            
            $channel->setAttribute('id', $newId);
            $channelMap[$oldId] = $newId;
        }
        
        // 處理所有節目節點
        $programmes = $xpath->query('//programme');
        foreach ($programmes as $programme) {
            $oldChannel = $programme->getAttribute('channel');
            if (isset($channelMap[$oldChannel])) {
                $programme->setAttribute('channel', $channelMap[$oldChannel]);
            }
        }
        
        return $dom->saveXML();
    } catch (Exception $e) {
        error_log("處理XML時發生錯誤: " . $e->getMessage());
        return null;
    }
}

/**
 * 處理單一來源
 */
function process_source($filename, $source) {
    $url = $source['url'];
    
    // 取得id_suffix，如果不存在則為空字串
    $idSuffix = isset($source['id_suffix']) ? $source['id_suffix'] : '';
    
    echo "開始處理來源: $filename" . 
         ($idSuffix ? " (ID後綴: $idSuffix)" : " (無ID後綴)") . "\n";
    
    // 下載XML內容
    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => $url,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT => 30,
        CURLOPT_SSL_VERIFYPEER => false,
        CURLOPT_FOLLOWLOCATION => true,
        CURLOPT_ENCODING => 'gzip'
    ]);
    
    $xmlContent = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);
    
    if ($httpCode !== 200 || empty($xmlContent)) {
        error_log("下載失敗: $filename (HTTP $httpCode)");
        return false;
    }
    
    // 處理XML - 使用id_suffix作為後綴（可能為空）
    $processedXml = process_xml($xmlContent, $idSuffix);
    if (!$processedXml) {
        error_log("處理失敗: $filename");
        return false;
    }
    
    // 儲存檔案 - 直接使用來源的鍵作為檔名
    $bytesWritten = file_put_contents($filename, $processedXml);
    if ($bytesWritten === false) {
        error_log("儲存失敗: $filename");
        return false;
    }
    
    echo "成功儲存: $filename (大小: " . number_format($bytesWritten) . " 位元組)\n";
    return true;
}

/**
 * 主處理函式（使用多進程）
 */
function main() {
    global $sources, $maxWorkers;
    
    $children = [];
    $results = [];
    
    foreach ($sources as $filename => $source) {
        $pid = pcntl_fork();
        
        if ($pid == -1) {
            die("無法建立子進程");
        } elseif ($pid) {
            // 父進程記錄PID
            $children[$pid] = $filename;
        } else {
            // 子進程處理
            $result = process_source($filename, $source);
            exit($result ? 0 : 1);
        }
        
        // 控制並行進程數
        while (count($children) > $maxWorkers) {
            $pid = pcntl_wait($status);
            if ($pid > 0) {
                $filename = $children[$pid];
                $results[$filename] = pcntl_wexitstatus($status) === 0;
                unset($children[$pid]);
            }
        }
    }
    
    // 等待剩餘子進程
    while (count($children) > 0) {
        $pid = pcntl_wait($status);
        if ($pid > 0) {
            $filename = $children[$pid];
            $results[$filename] = pcntl_wexitstatus($status) === 0;
            unset($children[$pid]);
        }
    }
    
    // 輸出結果
    $successCount = count(array_filter($results));
    $total = count($sources);
    echo "\n處理完成! 成功: $successCount/$total\n";
    
    // 列出所有產生的檔案
    echo "產生的檔案:\n";
    foreach (array_keys($sources) as $filename) {
        $status = isset($results[$filename]) && $results[$filename] ? "✓" : "✗";
        $size = file_exists($filename) ? number_format(filesize($filename)) . " 位元組" : "不存在";
        echo " - {$status} {$filename} ({$size})\n";
    }
    
    return $successCount === $total;
}

// 執行主程式
if (PHP_SAPI === 'cli') {
    if (!function_exists('pcntl_fork')) {
        die("需要啟用pcntl擴充功能\n");
    }
    
    // 確保當前目錄可寫入
    if (!is_writable('.')) {
        die("錯誤: 當前目錄不可寫入，請檢查權限\n");
    }
    
    main();
} else {
    die("請在CLI模式下執行此腳本\n");
}

File: epg_generator.py
#!/usr/bin/python3

import aiohttp
import asyncio
from lxml import etree
from concurrent.futures import ThreadPoolExecutor

# 待處理的EPG來源
sources = {
    "epgshanghai.xml": {
        "url": "https://epg.deny.vip/sh/tel-epg.xml",
        "source_id": "shanghai"  # 用於拼接頻道ID的原始標識
    },
    "epg51zmt.xml": {
        "url": "https://epg.51zmt.top:8001/difang.xml",
        "source_id": "51zmt"     # 用於拼接頻道ID的原始標識
    },
    "epgunifi.xml": {
        "url": "https://raw.githubusercontent.com/AqFad2811/epg/refs/heads/main/unifitv.xml",
        "source_id": ""     # 不存在或為空時，不拼接
    }
}

max_workers = 4    # 執行緒池大小

async def fetch_xml(session, url):
    """非同步取得XML內容"""
    try:
        async with session.get(url) as response:
            response.raise_for_status()
            return await response.text()
    except Exception as e:
        print(f"下載XML時發生錯誤: {str(e)}")
        return None

def get_display_name(channel):
    """
    智慧取得頻道顯示名稱，優先返回中文名稱
    處理邏輯：
    1. 優先尋找 lang="zh" 的 display-name
    2. 若無，則尋找任意 display-name（無論是否有 lang 屬性）
    3. 若都沒有，則返回 None
    """
    # 嘗試取得中文名稱
    zh_name_elem = channel.find(".//display-name[@lang='zh']")
    if zh_name_elem is not None and zh_name_elem.text:
        return zh_name_elem.text.strip()
    
    # 嘗試取得任意語言的名稱
    name_elems = channel.findall(".//display-name")
    for elem in name_elems:
        if elem.text and elem.text.strip():
            return elem.text.strip()
    
    # 嘗試從頻道ID取得
    channel_id = channel.get('id')
    if channel_id:
        return channel_id
    
    # 最終手段：返回預留位置名稱
    return "未知頻道"

def process_xml(xml_content, source_id):
    """使用lxml處理XML內容並格式化為標準XML"""
    if not xml_content:
        return None
    
    try:
        # 建立解析器，保留註解和CDATA
        parser = etree.XMLParser(
            remove_blank_text=True,  # 移除空白文字以便重新格式化
            remove_comments=False,   # 保留註解
            strip_cdata=False,       # 保留CDATA
            recover=True             # 嘗試復原錯誤
        )
        
        # 解析XML
        root = etree.fromstring(xml_content.encode('utf-8'), parser=parser)
        
        # 建立頻道ID映射表 (舊ID -> 新ID)
        channel_map = {}
        
        # 處理所有頻道節點
        for channel in root.xpath('.//channel'):
            channel_id = channel.get('id')
            
            # 智慧取得頻道名稱
            display_name = get_display_name(channel)
            
            if display_name:
                # 僅當source_id存在且非空時才拼接
                if source_id and source_id.strip():
                    new_id = f"{display_name}_{source_id}"
                else:
                    new_id = display_name
                
                # 更新頻道ID
                channel.set('id', new_id)
                channel_map[channel_id] = new_id
        
        # 處理所有節目節點
        for programme in root.xpath('.//programme'):
            old_channel = programme.get('channel')
            if old_channel in channel_map:
                programme.set('channel', channel_map[old_channel])
        
        # 建立新的XML樹以便格式化
        tree = etree.ElementTree(root)
        
        # 產生格式化後的XML字串
        xml_bytes = etree.tostring(
            tree,
            encoding='utf-8',
            xml_declaration=True,      # 包含XML宣告
            pretty_print=True,          # 啟用格式化輸出
            with_comments=True         # 保留註解
        )
        
        return xml_bytes.decode('utf-8')
    
    except Exception as e:
        print(f"處理XML時發生錯誤: {str(e)}")
        import traceback
        traceback.print_exc()
        return None

async def process_source(filename, source_info, session, executor):
    """處理單一來源：下載、處理並儲存XML"""
    url = source_info["url"]
    # 安全取得source_id，若不存在則為空字串
    source_id = source_info.get("source_id", "").strip()
    
    print(f"開始處理來源: {filename} (ID: '{source_id}')")
    
    # 非同步下載XML
    xml_content = await fetch_xml(session, url)
    if not xml_content:
        print(f"無法取得 {filename} 的XML內容")
        return False
    
    # 在執行緒池中同步處理XML（避免阻塞事件循環）
    processed_xml = await asyncio.get_event_loop().run_in_executor(
        executor, 
        process_xml, 
        xml_content, 
        source_id  # 傳遞原始標識用於拼接
    )
    
    if not processed_xml:
        print(f"處理 {filename} 的XML失敗")
        return False
    
    # 直接使用字典鍵作為檔名
    try:
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(processed_xml)
        print(f"成功儲存: {filename}")
        return True
    except Exception as e:
        print(f"儲存檔案 {filename} 時發生錯誤: {str(e)}")
        return False

async def main():
    """主非同步函式"""
    # 建立執行緒池用於處理CPU密集型任務
    executor = ThreadPoolExecutor(max_workers=max_workers)
    
    # 建立aiohttp會話
    async with aiohttp.ClientSession() as session:
        tasks = []
        for filename, source_info in sources.items():
            task = asyncio.create_task(
                process_source(filename, source_info, session, executor)
            )
            tasks.append(task)
        
        # 等待所有任務完成
        results = await asyncio.gather(*tasks)
        
        success_count = sum(1 for r in results if r)
        print(f"\n處理完成! 成功: {success_count}/{len(sources)}")
        
        # 關閉執行緒池
        executor.shutdown()

if __name__ == "__main__":
    asyncio.run(main())